### PR TITLE
Allow to tweak default scopes for accounts

### LINKED ISSUE #6582: Tweak the default scopes for Accounts
Currently the following default scopes for new accounts are defined inside `AccountManager`. We should let the server admin define them through the administration panel or config.php.
This is especially important on instances where users don't know each other.
```php
self::PROPERTY_DISPLAYNAME =>
    [
        'value' => $user->getDisplayName(),
        'scope' => self::VISIBILITY_CONTACTS_ONLY,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_ADDRESS =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_WEBSITE =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_EMAIL =>
    [
        'value' => $user->getEMailAddress(),
        'scope' => self::VISIBILITY_CONTACTS_ONLY,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_AVATAR =>
    [
        'scope' => self::VISIBILITY_CONTACTS_ONLY
    ],
self::PROPERTY_PHONE =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
self::PROPERTY_TWITTER =>
    [
        'value' => '',
        'scope' => self::VISIBILITY_PRIVATE,
        'verified' => self::NOT_VERIFIED,
    ],
```

### LINKED ISSUE #14358: set profile to private/public on new user creation
I'm always frustrated when I create multiple users and for them not to show up as contact in all other user contact lists, I have to log in as the new user and set the profile to private.

### Solution:

An option box to set the profile private in the user create dialogue

### Alternatives: 

1. set users profile to private per default, but this could lead to frustration if others want to have all users public per default
2. a setting in the config file that defines private or public for new users

### LINKED ISSUE #14959: don't reveal users mail adress by default
### Steps to reproduce
1. register as a new user
2. after some months, explore the menu
3. discover that your mail adress was public to ALL the other cloud users on the same server

### Expected behaviour
Private data should not be revealed **by default**
> Make public data available, protect private data.
Source: 
- hacker ethics
- gdpr

### Actual behaviour
Users mail adress is revealed to other users by default

### Server configuration

LAMP, Ubuntu 18.04

**Nextcloud version:** 15.0.5
fresh installl since upgrade from OC failed

**Where did you install Nextcloud from:** Webclient

**List of activated apps:**
deactivated ALL Apps - same behavior

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
If you have access to your command line run e.g.:

```{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.REMOVED.de"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "15.0.5.3",
        "overwrite.cli.url": "https:\/\/cloud.REMOVED.de",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "pipe",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "updater.secret": "***REMOVED SENSITIVE VALUE***",
        "maintenance": false,
        "theme": "",
        "loglevel": 2,
        "CUSTOMSETTINGSMATTHIAS": "ab hier!!!",
        "default_language": "de",
        "default_locale": "de",
        "force_language": "de",
        "force_locale": "de",
        "mail_smtpsecure": "ssl",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
    }
}```

### Client configuration
**Browser:**
IE, FF, Chrome
**Operating system:**

### PR DESCRIPTION
**Pull request by tcitworld**
Closes https://github.com/nextcloud/server/issues/6582, Closes https://github.com/nextcloud/server/issues/14358, Closes https://github.com/nextcloud/server/issues/14959

Replaces #20667

### CODE REVIEW

**Comment by szaimen** [comment]
I like this PR 👍
But I fear I am not the person to decide...

**Comment by szaimen** [comment]
@tcitworld can you please resolve the conflicts? Thank you :)

**Comment by szaimen** [comment]
Thanks! :)

File: config/config.sample.php [hunk 1]
```diff
@@ -2150,4 +2150,11 @@
  * the database storage.
  */
 'enable_file_metadata' => true,
+
+/**
+ * Allows to override the default scopes for Account data.
+ * The list of overridable properties and valid values for scopes are in
+ * OCP\Accounts\IAccountManager. Default values are in OC\Accounts\AccountManager
+ */
+'account_manager.default_property_scope' => []
```
Comment (juliushaertl) [inline config/config.sample.php:2159]: Would be nice to have an example in here directly on what to set and how. Could be in the comment of course to keep the actual value the default.

File: lib/public/Accounts/IAccountManager.php [hunk 2]
```diff
@@ -89,6 +90,21 @@ interface IAccountManager {
 	 */
 	public const VISIBILITY_PUBLIC = 'public';
 
+	/**
+	 * The list of allowed scopes
+	 *
+	 * @since 24.0.0
```
Comment (juliushaertl) [inline lib/public/Accounts/IAccountManager.php:96]:
```suggestion
	 * @since 25.0.0
```

File: lib/public/Accounts/IAccountManager.php [hunk 3]
```diff
@@ -122,6 +138,26 @@ interface IAccountManager {
 	 */
 	public const PROPERTY_PROFILE_ENABLED = 'profile_enabled';
 
+	/**
+	 * The list of allowed properties
+	 *
+	 * @since 24.0.0
```
Comment (juliushaertl) [inline lib/public/Accounts/IAccountManager.php:144]:
```suggestion
	 * @since 25.0.0
```

**Review by juliusknorr — COMMENTED** [review]
Minor comments, otherwise this looks good to me 👍

File: tests/lib/Accounts/AccountManagerTest.php [hunk 4]
```diff
@@ -431,10 +433,10 @@ protected function populateOrUpdate(): void {
 	/**
 	 * get a instance of the accountManager
 	 *
-	 * @param array $mockedMethods list of methods which should be mocked
+	 * @param array|null $mockedMethods list of methods which should be mocked
```
Comment (Pytal) [inline tests/lib/Accounts/AccountManagerTest.php:436]:
```suggestion
```

File: tests/lib/Accounts/AccountManagerTest.php [hunk 5]
```diff
@@ -431,10 +433,10 @@ protected function populateOrUpdate(): void {
...
 	 * @return MockObject | AccountManager
 	 */
-	public function getInstance($mockedMethods = null) {
+	public function getInstance(array $mockedMethods = null) {
```
Comment (Pytal) [inline tests/lib/Accounts/AccountManagerTest.php:439]:
```suggestion
	public function getInstance(?array $mockedMethods = null) {
```

**Review by Pytal — COMMENTED** [review]
🔭

File: config/config.sample.php (hunk 1 above)
Comment (tcitworld) [reply config/config.sample.php:2182]: Added

**Comment by Pytal** [comment]
Could we prevent erroneously allowing displayname and email to be set to private in the logic? Unsupported as per https://github.com/nextcloud/server/blob/778db45631929d7caa07d3602ba7f462f0908c4b/lib/private/Accounts/AccountManager.php#L218-L228
In these cases we should default to local scope

**Comment by jancborchardt** [comment]
@tcitworld can you add before/after screenshots for easier design review? Thanks! :)

**Comment by tcitworld** [comment]
> Could we prevent erroneously allowing displayname and email to be set to private in the logic? 

Done!

> can you add before/after screenshots for easier design review?

I really can't ! 😅
This is a new advanced setting for administrators through `config.php` only.

**Review by Pytal — APPROVED** [review]

**Review by juliusknorr — APPROVED** [review]

**Comment by juliushaertl** [comment]
Failures are unrelated.

**Comment by szaimen** [comment]
Should we add some documentation on this with a complete example?

**Comment by tcitworld** [comment]
Yup, needs an entry in https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html

**Comment by tcitworld** [comment]
Yes, more appropriate. But still needs a quick mention just before https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#app-config-options

**Comment by szaimen** [comment]
> But still needs a quick mention just before https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#app-config-options

If I understood this section in the settings documentation correctly, it is actually auto-generated. So I think we should not need to edit this section manually? Maybe @artonge knows?

**Comment by szaimen** [comment]
But of course the profile_configuration section should then be edited and include this setting, I suppose :)

**Comment by tcitworld** [comment]
Oh right, I missed that !

**Comment by szaimen** [comment]
@tcitworld can you open a PR with some documentation on this over at https://github.com/nextcloud/documentation/edit/master/admin_manual/configuration_user/profile_configuration.rst or shall I try to add some documentation on this? :)

**Comment by szaimen** [comment]
Thank you! 🚀

**Pull request merged**